### DISCUSSION TITLE
Convergence behavior

### DISCUSSION THREAD

**Discussion by SchroederSa** [svenreiche/Genesis-1.3-Version4#199] [Q&A] [answered]
Hi,

I would like to understand the convergence behaviour of my simulations. As such, I was scanning parameters that seemed to me most relevant:

'time/slen'
'field/ngrid'
'field/dgrid'
'setup/nbins' & 'setup/npart'

wit the following input:

```
&setup
rootname=nominal
outputdir=outputdir
lattice=lattice.lat
beamline=BL
lambda0=421e-9
gamma0=195.7
delz=0.054
seed=1
nbins=4
npart=16384
shotnoise=1
field_global_stat=true
beam_global_stat=true
write_meta_file=false
exclude_spatial_output=true
exclude_fft_output=true
exclude_current_output=true
exclude_aux_output=true
&end

&lattice
zmatch=0.252
&end

&time
s0=-50e-6
slen=150e-6
sample=3
&end

&field
power=0
ngrid=2401
dgrid=2.5e-3
waist_size=205.18e-6
&end

&profile_gauss
label=current
c0=1193
s0=0
sig=10e-6
&end

&profile_polynom
label=energy
c0=195.7
c1=985000
&end

&beam
gamma=195.7
current=@current
delgam=0.56
ex=1e-6
ey=1e-6
&end

&track
&end
```
However, I am a bit puzzled by the results, is this an expected level of convergence? The seed is kept always the same. And there are clear convergence tendencies in some parameters (```dgrid```) but non in others (```number of beamlets = npart/nbins```). Also for ```ngrid```, the energy is well converged but the far field intensity is not. Am I missing a factorisation here?

The solid line is the energy: ```hid['Field']['Global']['energy'][()]*1e6```
The dashed line the farfield intensity: ```hid['Field']['Global']['intensity-farfield'][()]*1e-21```

<img width="662" alt="Screenshot 2024-09-30 at 12 22 49 PM" src="https://github.com/user-attachments/assets/c12407b3-01b2-42f2-887f-ab7e7d1c66f9">
<img width="659" alt="Screenshot 2024-09-30 at 12 22 57 PM" src="https://github.com/user-attachments/assets/9c021d90-abe1-43d3-9c52-94c89b32d684">
<img width="653" alt="Screenshot 2024-09-30 at 12 23 06 PM" src="https://github.com/user-attachments/assets/f2c10d11-9200-403b-af84-d8ea997129f2">
<img width="657" alt="Screenshot 2024-09-30 at 12 23 19 PM" src="https://github.com/user-attachments/assets/41eb7414-37a9-4572-8aaa-f675dd04bfc1">
<img width="660" alt="Screenshot 2024-09-30 at 12 23 27 PM" src="https://github.com/user-attachments/assets/4e570b9d-60b6-479d-9fc9-b7e7ec6491f0">

Many thanks for your help and advice!
Sarah

**Comment by svenreiche** [comment]
Hi Sarah,
some can be explained, some others not so much. But let me reproduce the scan on my clusters to exclude some possibilities.
Also the number of grid points is very high. This cause that the simulation is getting much more sensitive to higher modes.
I will report back once I have done the convergence studies for your input deck.
Sven

**Comment by svenreiche** [comment, marked as answer]
Ok. It was faster than I though (on my cluster the jobs are taking about 2 min for 360 slices (I set sample to 1) using 36 processes.

For slen, nbins and npart the problem lies in the number generator. Despite having the same global seed, the cases have different times of polling a random number.  Whenever the number of npart changes then there is no real control on what a given slice should produce the same bunching factor. Since one adds up N random number or 2*N random number. To have consistencies it would beam that the second half should have a perfect mean value, which is an additional requirement which is not needed.
With slen it is a little bit more subtle but at the end it is the same. For multi-core simulation the global seed is transferred to each core and then added its rank to it to initialized the sequence. Since the slices per process changes the random number sequence is shifted, e.g. for slen = 150 microns, one core might hold 10 slices. So the process with rank =1 initialized the 11th slices. Now if slen is changed so is the number of slices. Now the process with rank =1 starts at the tenth slice. Thus the random number profile is shifted in time and thus not comparable to the other runs.

Note that SEED guarantees a reproducible results if the same input deck is run again under the same conditions. Anything else will change the results.

Since this simulation is quasi a single spike acting in weak superradiance the physical fluctuation is tremendous up to saturation where the difference is not that big when compared to the linear regime. This is the nature of the SASE process.  For longer pulses with many spikes the difference will average out.  **So for your input deck it is almost impossible to separate the numerical fluctuation from the physical one**, unless one runs about 100 cases for the same settings except different seeds.

About the dependence of the far field on ngrid.  Increasing the number of grid points allows for more transverse modes, in particular those with a larger divergence (here the maximum divergence angle scales with the grid spacing between two adjacent grid points). It is fairly easy to have some divergence which violates the solution of the paraxial equation for the field solver. As a results there are some alias effects when the higher modes bounces back on the grid. This can artificially contribute to the 'seen' on-axis field. Therefore there is a systematic coupling of higher transverse mode to the fundamental on axis mode.  Therefore its absolute value has to be taken with a grain of salt. TO be more consistent it would be better to normalize the farfield-intensity to the radiation energy, which has no dependence on the grid points (since all relevant FEL physics is done with a really few transverse modes).

I hope this makes sense or is at least digestible.

Best
Sven

**Comment by SchroederSa** [comment]
Thank you for your clear and detailed reply and explanation - I believe I grasped most of it. Very insightful! I didn't expect to learn that much by looking at convergence tests. 
My latest tests look very convincing!

<img width="659" alt="Screenshot 2024-10-01 at 8 51 36 PM" src="https://github.com/user-attachments/assets/a4ab8ebe-ae19-40ac-8b86-455b2f50f63d">
<img width="653" alt="Screenshot 2024-10-01 at 8 51 42 PM" src="https://github.com/user-attachments/assets/a03a3e72-d395-4edc-8e81-a97ea40533c9">
<img width="645" alt="Screenshot 2024-10-01 at 8 51 49 PM" src="https://github.com/user-attachments/assets/68f1e9c0-cbb0-42a9-a5fa-d2b24e7884e1">
<img width="670" alt="Screenshot 2024-10-01 at 8 51 56 PM" src="https://github.com/user-attachments/assets/4034d6f1-2aee-4b92-a027-f9ead641f79d">
<img width="651" alt="Screenshot 2024-10-01 at 8 52 04 PM" src="https://github.com/user-attachments/assets/23a3c5f5-238a-4435-90d6-cfea7aa3cfb4">
<img width="648" alt="Screenshot 2024-10-01 at 8 52 10 PM" src="https://github.com/user-attachments/assets/d57fab81-b744-4110-8513-d435a655d932">
<img width="639" alt="Screenshot 2024-10-01 at 8 52 23 PM" src="https://github.com/user-attachments/assets/44a6359e-9928-41a8-8484-fdbb3ce34cb7">
<img width="639" alt="Screenshot 2024-10-01 at 8 52 37 PM" src="https://github.com/user-attachments/assets/279d4ee5-d3c5-41a9-b204-79d3024ce41e">
